File: tron-backend/src/worker.js
require('dotenv').config();
const Redis = require('ioredis');
const { supabase } = require('./config/supabase.js');
const PMOrchestrator = require('./adapters/pm-orchestrator');
const githubAdapter = require('./adapters/github'); 
const aiAdapter = require('./adapters/ai');
const messengerAdapter = require('./adapters/messenger');
const redis = require('./config/redis');
const { getRepoConfigFromDB } = require('./config/db.js');

// 🛡️ THE FIX: Smart connection that won't freeze silently
const redisUrl = process.env.REDIS_URL || 'redis://127.0.0.1:6379';
const isSecure = redisUrl.startsWith('rediss://'); 


// Add logs so we can physically see the worker connect!
redis.on('connect', () => console.log('✅ [Worker] Successfully connected to Redis!'));
redis.on('error', (err) => console.error('❌ [Worker] Redis Connection Error:', err.message));

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

console.log('👷 T.R.O.N. V3 Background Worker Booting Up...');

async function startWorker() {
    console.log('🎧 Worker is actively listening to the Redis Queue...');

    while (true) {
        try {
            const jobString = await redis.rpop('tron:v3_secret_queue');
            
            if (!jobString) {
                await sleep(1500);
                continue;
            }

            console.log(`\n🔥 [WORKER] DATA FOUND!`);
            const job = JSON.parse(jobString);

            console.log(`⚙️  Processing Delivery ID: [${job.deliveryId}]`);

            // 3. Extract repo name
            const repoName = job.payload.repository?.full_name;
            if (!repoName) {
                console.log('⏭️  Skipping: Payload does not contain a repository name.');
                continue;
            }

            // 🧠 THE FIX: Asynchronously fetch the config from Supabase!
            // 🌟 Here is our true variable name: projectConfig
            const projectConfig = await getRepoConfigFromDB(repoName);
            
            if (!projectConfig) {
                console.log(`⏭️  Skipping: No configuration found in Database for repo "${repoName}"`);
                continue;
            }

            // ==========================================
            // EVENT: PULL REQUEST
            // ==========================================
            if (job.eventType === 'pull_request') {
                const prTitle = job.payload.pull_request.title;
                const branchName = job.payload.pull_request.head.ref || "";
                const action = job.payload.action; 
                
                console.log(`\n🔀 [PR EVENT] Action: ${action} | Title: "${prTitle}"`);

                let taskIdentifier = null;
                const branchMatch = branchName.match(/(\d{9,}|[A-Z]+-\d+)/); 
                const titleMatch = prTitle.match(/(\d{9,}|[A-Z]+-\d+)/);

                if (branchMatch) {
                    taskIdentifier = branchMatch[1];
                    console.log(`🎯 Extracted Task ID [${taskIdentifier}] from PR branch name.`);
                } else if (titleMatch) {
                    taskIdentifier = titleMatch[1];
                    console.log(`🎯 Extracted Task ID [${taskIdentifier}] from PR title.`);
                }

                // --- PHASE 1: PM STATE TRACKING ---
                const pmTool = {
                    ...projectConfig.pm_tool,
                    project_id: projectConfig.pm_tool?.board_id || projectConfig.pm_tool?.project_id || projectConfig.pm_project_id
                };

                const mappingKey = `pull_request_${action}`; 
                const newStatus = projectConfig.mapping[mappingKey];

                if (!newStatus) {
                    console.log(`⏭️  Skipping PM update: No mapping found in DB for "${mappingKey}"`);
                } else if (taskIdentifier && pmTool && pmTool.provider !== "none") {
                    try {
                        console.log(`🚚 Moving ticket [${taskIdentifier}] to column ${newStatus} in ${pmTool.provider}...`);
                        await PMOrchestrator.updateTicketStatus(pmTool, taskIdentifier, newStatus);
                        console.log(`✅ Successfully moved ticket for PR ${action}!`);
                    } catch (error) {
                        console.error(`⚠️ Failed to move PM ticket:`, error.message);
                    }
                }

                // --- PHASE 2: AI PIPELINE (DIFF SANITIZER & SUMMARIZATION) ---
                if (action === 'opened' || action === 'reopened') {
                    if (job.payload.pull_request.draft === true) {
                        console.log(`⏭️  [AI PIPELINE] Skipping Draft PR: "${prTitle}"`);
                        continue; 
                    }

                    const diffUrl = job.payload.pull_request.diff_url;
                    const repoFullName = job.payload.repository.full_name; 
                    const prNumber = job.payload.pull_request.number;

                    console.log(`\n🧠 [AI PIPELINE] Generating Intel & Code Review for: "${prTitle}"`);
                    
                    try {
                        const sanitizedDiff = await githubAdapter.fetchAndSanitizeDiff(diffUrl);
                        
                        console.log(`🕵️‍♂️ Analyzing diff for bugs...`);
                        const codeReview = await aiAdapter.generateCodeReview(sanitizedDiff);
                        
                        // Store the review in Redis for 7 days (604800 seconds)
                        if (taskIdentifier) {
                            await redis.set(`ai_review:${taskIdentifier}`, codeReview, 'EX', 604800);
                            console.log(`💾 Saved AI Code Review to Redis for Task [${taskIdentifier}]`);
                        }
                        
                        console.log(`💬 Posting Code Review to GitHub PR #${prNumber}...`);
                        const commentHeader = `### 🤖 T.R.O.N. Automated Code Review\n\n`;
                        await githubAdapter.postPullRequestComment(repoFullName, prNumber, commentHeader + codeReview);

                        const intelligenceReport = await aiAdapter.generateExecutiveSummary(prTitle, sanitizedDiff);
                        
                        console.log(`\n📊 --- FINAL EXECUTIVE REPORT ---`);
                        console.log(`🏷️  Category: ${intelligenceReport.intent}`);
                        console.log(`📝 Summary:  ${intelligenceReport.executive_summary}`);
                        console.log(`🚀 Impact:   ${intelligenceReport.business_impact}`);
                        console.log(`--------------------------------\n`);

                        const prUrl = job.payload.pull_request.html_url;

                        if (projectConfig.communication) {
                            await messengerAdapter.broadcastSummary(projectConfig.communication, prTitle, prUrl, intelligenceReport);
                        } else {
                            console.log(`⚠️ No communication config found in DB. Skipping broadcast.`);
                        }

                    } catch (aiError) {
                        console.error(`❌ [AI PIPELINE] Pipeline failed:`, aiError.message);
                    }
                }

            // ==========================================
            // EVENT: LOCAL DAEMON TASK START
            // ==========================================
            } else if (job.eventType === 'local_start') {
                const taskID = job.payload.taskId;
                
                const pmTool = {
                    ...projectConfig.pm_tool,
                    project_id: projectConfig.pm_tool?.board_id || projectConfig.pm_tool?.project_id || projectConfig.pm_project_id
                };
                
                const newStatus = projectConfig.mapping['branch_created']; 

                if (!newStatus) {
                    console.log(`⏭️  Skipping: No 'branch_created' mapping found in DB`);
                    continue;
                }

                try {
                    console.log(`🚚 Moving ticket [${taskID}] to ${newStatus}...`);
                    await PMOrchestrator.updateTicketStatus(pmTool, taskID, newStatus);
                    console.log(`✅ Successfully moved ticket [${taskID}]!`); 
                } 
                catch (error){
                    console.error(`❌ [ORCHESTRATOR] Failed to update status:`, error.message);
                    
                    job.retryCount = (job.retryCount || 0) + 1;
                    
                    if (job.retryCount <= 3) {
                        const backoffTime = Math.pow(2, job.retryCount) * 1000;
                        console.log(`⏳ API Overloaded. Applying backoff. Waiting ${backoffTime}ms...`);
                        await sleep(backoffTime);
                        await redis.lpush('tron:v3_secret_queue', JSON.stringify(job)); // FIXED QUEUE NAME
                    } else {
                        console.error(`💀 Job permanently failed. Moving to Dead Letter Queue.`);
                        await redis.lpush('tron:dead_letters', jobString);
                    }
                    continue; 
                }

            // ==========================================
            // EVENT: GITHUB PUSH / BRANCH CREATED
            // ==========================================
            } else if (job.eventType === 'push' || job.eventType === 'create') {
                const ref = job.payload.ref || "";
                
                if (job.payload.deleted === true) {
                    console.log(`🗑️  [BRANCH EVENT] Branch deleted. Ignoring to prevent ticket rewind.`);
                    continue; 
                }

                if (job.eventType === 'push' && !ref.startsWith('refs/heads/')) {
                    console.log(`⏭️  Skipping non-branch push: ${ref}`);
                } else {
                    const branchName = ref.replace('refs/heads/', '');
                    console.log(`\n🌿 [BRANCH EVENT] Detected branch: "${branchName}"`);

                    const taskIdMatch = branchName.match(/(\d{9,}|[A-Z]+-\d+)/); 

                    if (taskIdMatch) {
                        const taskIdentifier = taskIdMatch[1];
                        console.log(`🎯 Extracted Task ID [${taskIdentifier}] from branch name.`);

                        const pmTool = {
                            ...projectConfig.pm_tool,
                            project_id: projectConfig.pm_tool?.board_id || projectConfig.pm_tool?.project_id || projectConfig.pm_project_id
                        };
                        const newStatus = projectConfig.mapping['branch_created']; 

                        if (!newStatus) {
                            console.log(`⏭️  Skipping: No 'branch_created' mapping found in DB`);
                        } else if (pmTool && pmTool.provider !== "none") {
                            try {
                                console.log(`🚚 Moving ticket [${taskIdentifier}] to ${newStatus}...`);
                                await PMOrchestrator.updateTicketStatus(pmTool, taskIdentifier, newStatus);
                                console.log(`✅ Successfully moved ticket!`);
                            } catch (error) {
                                console.error(`⚠️ Failed to move PM ticket:`, error.message);
                            }
                        }
                    } else {
                        console.log(`⚠️  No valid Task ID found in branch name "${branchName}". Cannot move PM card.`);
                    }
                }
            }

        } catch (error) {
            console.error('❌ Critical Worker Error:', error);
            await sleep(5000); // Prevent tight loop crashing
        }
    }
}

// Wait for the secure SSL connection to finish before pulling jobs
redis.once('ready', () => {
    startWorker();
});

process.on('SIGTERM', async () => {
    console.log('\n🛑 SIGTERM received. Shutting down worker gracefully...');
    await redis.quit();
    console.log('💤 Disconnected from Redis. Exiting process.');
    process.exit(0);
});

process.on('SIGINT', async () => {
    console.log('\n🛑 SIGINT received. Shutting down worker gracefully...');
    await redis.quit();
    console.log('💤 Disconnected from Redis. Exiting process.');
    process.exit(0);
});

File: tron-go-backend/internal/worker/worker.go
package worker

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/tron-v3.1/tron-go-backend/internal/adapters"
	"github.com/tron-v3.1/tron-go-backend/internal/models"
	"github.com/tron-v3.1/tron-go-backend/internal/services"
	"github.com/tron-v3.1/tron-go-backend/pkg/database"
	"github.com/tron-v3.1/tron-go-backend/pkg/logger"
	"github.com/tron-v3.1/tron-go-backend/pkg/redis"
)

type QueueJob struct {
	DeliveryID string                 `json:"deliveryId"`
	EventType  string                 `json:"eventType"`
	Payload    map[string]interface{} `json:"payload"`
}

// Start begins the Redis polling loop in the background
func Start(ctx context.Context) {
	logger.Log.Info("📡 Integrated T.R.O.N. Worker is live and listening to 'tron:v3_secret_queue'")

	githubAPI := adapters.NewGitHubAdapter()
	basecampAPI := adapters.NewBasecampAdapter()
	aiAPI := adapters.NewAIAdapter()
	messengerAPI := adapters.NewMessengerAdapter()
	orchestrator := services.NewPMOrchestrator(basecampAPI)

	for {
		select {
		case <-ctx.Done():
			logger.Log.Info("🛑 Halting integrated worker queue polling...")
			return
		default:
			result, err := redis.Client.BRPop(ctx, 2*time.Second, "tron:v3_secret_queue").Result()

			if err == context.Canceled {
				return
			}
			if err != nil {
				if err.Error() != "redis: nil" {
					logger.Log.Warnf("⚠️ Redis pop error: %v", err)
					time.Sleep(1 * time.Second)
				}
				continue
			}

			var job QueueJob
			if err := json.Unmarshal([]byte(result[1]), &job); err != nil {
				logger.Log.Errorf("❌ Failed to parse job JSON: %v", err)
				continue
			}

			go processJob(job, githubAPI, aiAPI, messengerAPI, orchestrator)
		}
	}
}

func processJob(job QueueJob, githubAPI *adapters.GitHubAdapter, aiAPI *adapters.AIAdapter, messengerAPI *adapters.MessengerAdapter, orchestrator *services.PMOrchestrator) {
	logger.Log.Infof("\n⚙️  [WORKER] Processing Job [%s] - Type: %s\n", job.DeliveryID, job.EventType)

	// 🌟 FIX 1: Handle local VS Code task starts gracefully without dropping them
	if job.EventType == "local_start" {
		logger.Log.Infof("💻 [WORKER] Developer started task locally: %v\n", job.Payload["taskId"])
		// Future expansion: Add Discord broadcast logic here (e.g., "Omkar started working on TRON-123")
		return
	}

	if job.EventType != "pull_request" {
		return
	}

	action, _ := job.Payload["action"].(string)

	switch action {
	case "opened", "reopened":
		handleNewPullRequest(job.Payload, githubAPI, aiAPI, messengerAPI, orchestrator)
	case "closed":
		handleClosedPullRequest(job.Payload, orchestrator)
	default:
		// ignore other actions
	}
}

func handleClosedPullRequest(payload map[string]interface{}, orchestrator *services.PMOrchestrator) {
	prData, _ := payload["pull_request"].(map[string]interface{})
	repoData, _ := payload["repository"].(map[string]interface{})

	repoFullName, _ := repoData["full_name"].(string)
	prTitle, _ := prData["title"].(string)

	prNumber := 0
	if num, ok := prData["number"].(float64); ok {
		prNumber = int(num)
	}

	logger.Log.Infof("📦 [PIPELINE] PR Closed Detected: %s (#%d)\n", repoFullName, prNumber)

	var repoConfig models.Repository
	if err := database.DB.Where("repo_name = ?", repoFullName).First(&repoConfig).Error; err != nil {
		logger.Log.Warnf("⚠️ [PIPELINE] Repo '%s' is not mapped in TRON. Ignoring.\n", repoFullName)
		return
	}
	orgID := repoConfig.OrgID

	var mapping map[string]interface{}
	json.Unmarshal([]byte(repoConfig.Mapping), &mapping)

	headData, _ := prData["head"].(map[string]interface{})
	branchName, _ := headData["ref"].(string)

	searchQuery := prTitle
	if branchName != "" {
		searchQuery = prTitle + " " + branchName
	}

	_, exactUrl := orchestrator.ResolveTask(repoConfig.PMProvider, repoConfig.PMProjectID, searchQuery, orgID, mapping)

	if exactUrl != "" {
		extractID := func(key string) string {
			var target map[string]interface{} = mapping
			if nested, ok := mapping["mapping"].(map[string]interface{}); ok {
				target = nested
			}

			if val, ok := target[key].(string); ok {
				return val
			} else if val, ok := target[key].(float64); ok {
				return fmt.Sprintf("%.0f", val)
			}
			return ""
		}

		doneCol := extractID("pull_request_closed")
		if doneCol == "" {
			doneCol = extractID("done")
		}

		if doneCol != "" {
			logger.Log.Infof("🚚 [PIPELINE] Moving task to Done column (%s)...\n", doneCol)
			orchestrator.UpdateTicketStatus(repoConfig.PMProvider, repoConfig.PMProjectID, exactUrl, doneCol, orgID)
		} else {
			logger.Log.Warnf("⚠️ [PIPELINE] Could not find 'pull_request_closed' in mapping. Skipping column move.\n")
		}
	}
}

func handleNewPullRequest(payload map[string]interface{}, githubAPI *adapters.GitHubAdapter, aiAPI *adapters.AIAdapter, messengerAPI *adapters.MessengerAdapter, orchestrator *services.PMOrchestrator) {
	prData, _ := payload["pull_request"].(map[string]interface{})
	repoData, _ := payload["repository"].(map[string]interface{})
	senderData, _ := payload["sender"].(map[string]interface{})

	repoFullName, _ := repoData["full_name"].(string)

	prNumber := 0
	if num, ok := prData["number"].(float64); ok {
		prNumber = int(num)
	}

	prTitle, _ := prData["title"].(string)
	developerName, _ := senderData["login"].(string)

	var repoConfig models.Repository
	if err := database.DB.Where("repo_name = ?", repoFullName).First(&repoConfig).Error; err != nil {
		logger.Log.Warnf("⚠️ [PIPELINE] Repo '%s' is not mapped in TRON. Ignoring.\n", repoFullName)
		return
	}
	orgID := repoConfig.OrgID

	var installID string
	if installData, ok := payload["installation"].(map[string]interface{}); ok && installData != nil {
		if idFloat, ok := installData["id"].(float64); ok {
			installID = fmt.Sprintf("%.0f", idFloat)
		} else if installData["id"] != nil {
			installID = fmt.Sprintf("%v", installData["id"])
		}
	}

	if installID == "" || installID == "<nil>" {
		var githubInt models.Integration
		if err := database.DB.Where("org_id = ? AND provider = ?", orgID, "github").First(&githubInt).Error; err == nil {
			installID = githubInt.Token
			if installID == "" && githubInt.SecretID != nil {
				installID, _ = services.GetDecryptedSecret(*githubInt.SecretID)
			}
		}
	}

	if installID == "" || installID == "<nil>" {
		logger.Log.Errorf("❌ [PIPELINE] Could not find GitHub Installation ID in payload or Database. Aborting.\n")
		return
	}

	logger.Log.Infof("📦 [PIPELINE] New PR Detected: %s (#%d) by %s (Install ID: %s)\n", repoFullName, prNumber, developerName, installID)

	sanitizedDiff, err := githubAPI.FetchAndSanitizeDiff(repoFullName, prNumber, installID)
	if err != nil {
		logger.Log.Errorf("❌ [PIPELINE] Failed to fetch diff: %v\n", err)
		return
	}

	logger.Log.Info("🧠 [PIPELINE] Analyzing Code...")
	aiSummary, _ := aiAPI.GenerateExecutiveSummary(prTitle, sanitizedDiff)
	aiCodeReview := aiAPI.GenerateCodeReview(sanitizedDiff)

	githubAPI.PostPullRequestComment(repoFullName, prNumber, aiCodeReview, installID)

	logger.Log.Info("🏗️  [PIPELINE] Synchronizing Project Management Boards...")
	var mapping map[string]interface{}
	json.Unmarshal([]byte(repoConfig.Mapping), &mapping)

	headData, _ := prData["head"].(map[string]interface{})
	branchName, _ := headData["ref"].(string)

	searchQuery := prTitle
	if branchName != "" {
		searchQuery = prTitle + " " + branchName
	}

	_, exactUrl := orchestrator.ResolveTask(repoConfig.PMProvider, repoConfig.PMProjectID, searchQuery, orgID, mapping)

	if exactUrl != "" {
		// Assign Developer
		orchestrator.AssignTicket(repoConfig.PMProvider, repoConfig.PMProjectID, exactUrl, developerName, orgID)

		// 🌟 FIX 2: Save AI Review to Redis for the VS Code Extension (Expires in 7 Days)
		err := redis.Client.Set(context.Background(), "ai_review:"+exactUrl, aiCodeReview, 7*24*time.Hour).Err()
		if err != nil {
			logger.Log.Warnf("⚠️ [PIPELINE] Failed to cache AI review in Redis: %v\n", err)
		} else {
			logger.Log.Infof("💾 [PIPELINE] Cached AI Review in Redis for VS Code (Key: ai_review:%s)\n", exactUrl)
		}

		extractID := func(key string) string {
			var target map[string]interface{} = mapping
			if nested, ok := mapping["mapping"].(map[string]interface{}); ok {
				target = nested
			}

			if val, ok := target[key].(string); ok {
				return val
			} else if val, ok := target[key].(float64); ok {
				return fmt.Sprintf("%.0f", val)
			}
			return ""
		}

		underReviewCol := extractID("pull_request_opened")
		if underReviewCol == "" {
			underReviewCol = extractID("pr_opened")
		}
		if underReviewCol == "" {
			underReviewCol = extractID("under_review")
		}

		if underReviewCol != "" {
			logger.Log.Infof("🚚 [PIPELINE] Moving task to PR/Under Review column (%s)...\n", underReviewCol)
			orchestrator.UpdateTicketStatus(repoConfig.PMProvider, repoConfig.PMProjectID, exactUrl, underReviewCol, orgID)
		} else {
			logger.Log.Warnf("⚠️ [PIPELINE] Could not find 'pr_opened' or 'under_review' in mapping. Skipping column move.\n")
		}
	}

	logger.Log.Info("📢 [PIPELINE] Broadcasting to team...")
	var commConfig adapters.CommunicationConfig
	json.Unmarshal([]byte(repoConfig.CommunicationConfig), &commConfig)

	report := adapters.AIReport{
		Intent:           aiSummary.Intent,
		ExecutiveSummary: aiSummary.ExecutiveSummary,
		BusinessImpact:   aiSummary.BusinessImpact,
	}

	prURL, _ := prData["html_url"].(string)
	messengerAPI.BroadcastSummary(commConfig, prTitle, prURL, report, orgID)

	logger.Log.Infof("🏁 [PIPELINE] Successfully processed PR #%d\n", prNumber)
}
